### PR TITLE
[DEV-1115]: Solana sales integration

### PR DESCRIPTION
**Pull request by Superjo149**
- Use a combination of metaplex/js and code from the repositories to be able to buy and sell
- Added Transfer & Burn
- Created an abstraction to instantiate the RefinableClient for solana and evm
- Created an abstraction to wrap both Solana and Evm transactions
- Added unit tests & run tests in CI

Created a new next branch to merge to, so that we can start working on that without interrupting and putting the new package version up as latest until everything is tested further.

### CODE REVIEW

File: src/offer/SaleOffer.ts [hunk 1]
```diff
@@ -1,39 +1,42 @@
...
+      ownerEthAddress: this.user?.ethAddress,
+      royaltyContractAddress: params.royaltyContractAddress,
       supply,
-      amount
-    );
+      amount,
+      blockchainId: this.blockchainId,
+    });
 
     await this.refinable.apiClient.request<
```
Comment (viet10101988) [inline src/offer/SaleOffer.ts:41]: can the this.nft.buy() before fail? If so need to adjust this call accordingly

File: src/refinable/account/SolanaAccount.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  /**
+   * Balance of Any Token (converted from wei).
+   * @return {Promise<string>}
+   */
+  public async getTokenBalance(tokenAddress: string): Promise<string> {
+    throw new Error("Not implemented");
+  }
+
+  public async getBalance(): Promise<string> {
+    const result = await this.refinable.connection.getBalance(
```
Comment (viet10101988) [inline src/refinable/account/SolanaAccount.ts:21]: could this fail? Might break other places if they use getBalance()

File: src/transaction/EvmTransaction.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  get txId() {
+    return this.tx.hash;
+  }
+
+  get timestamp() {
+    return this.tx.timestamp;
+  }
+
+  get success() {
+    if (!this.txReceipt) throw new Error("Must await tx first");
```
Comment (viet10101988) [inline src/transaction/EvmTransaction.ts:20]: do we need throw if this is just a get()?

File: src/transaction/SolanaTransaction.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  timestamp: number;
+
+  constructor(
+    public readonly txId: string,
+    private readonly connection: Connection
+  ) {}
+
+  get success() {
+    if (!this.txResult) throw new Error("Must await tx first");
```
Comment (viet10101988) [inline src/transaction/SolanaTransaction.ts:20]: same as above

**Review by viet10101988 — CHANGES_REQUESTED** [review]

File: examples/item/itemsOnAuction.ts [hunk 5]
```diff
@@ -10,7 +9,7 @@ async function main() {
   const chainId = Chain.BscTestnet;
   const wallet = initializeWallet(PRIVATE_KEY, chainId);
   try {
-    const refinable = await Refinable.create(wallet, API_KEY, {
+    const refinable = await RefinableEvmClient.create(wallet, API_KEY, {
```
Comment (maximgeerinck) [inline examples/item/itemsOnAuction.ts:12]: not sure why we didn't put this in a factory with a network selector

File: src/config/sdk.ts [hunk 6]
```diff
@@ -39,12 +39,8 @@ export const getContractsTags = (
       ];
     case Environment.Local:
       return [
-        chain === ChainType.BSC
-          ? ContractTag.SaleV4_0_0
-          : ContractTag.SaleV3_0_1,
-        chain === ChainType.BSC
-          ? ContractTag.AuctionV4_0_0
-          : ContractTag.AuctionV3_1_0,
```
Comment (maximgeerinck) [inline src/config/sdk.ts:47]: why was the v4 removed? it's the diamond contract

File: src/solana/actions/addTokensToVault.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  const vaultAuthority = (
+    await findProgramAddress(
+      [
+        Buffer.from(VAULT_PREFIX),
+        toPublicKey(PROGRAM_IDS.vault).toBuffer(),
+        toPublicKey(vault).toBuffer(),
+      ],
+      toPublicKey(PROGRAM_IDS.vault)
+    )
+  )[0];
```
Comment (maximgeerinck) [inline src/solana/actions/addTokensToVault.ts:61]: potential error ([0])

File: src/solana/constants/labels.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    'This page was produced by the Solana Foundation ("SF") for internal educational and inspiration purposes only. SF does not encourage, induce or sanction the deployment, integration or use of Oyster or any similar application (including its code) in violation of applicable laws or regulations and hereby prohibits any such deployment, integration or use. Anyone using this code or a derivation thereof must comply with applicable laws and regulations when releasing related software.',
+  MENU_HOME: 'Home',
+  MENU_DASHBOARD: 'Dashboard',
+  CONNECT_BUTTON: 'Connect',
+  WALLET_TOOLTIP: 'Wallet public key',
+  WALLET_BALANCE: 'Wallet balance',
+  SETTINGS_TOOLTIP: 'Settings',
+  DASHBOARD_ACTION: 'Go to dashboard',
+  GO_BACK_ACTION: 'Go back',
+};
```
Comment (maximgeerinck) [inline src/solana/constants/labels.ts:15]: do we need these?

File: src/utils/connection.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { Connection } from "@solana/web3.js";
+import { Chain } from "..";
+import { getChainByNetworkId } from "../config/chains";
+
+export const getConnectionByChainId = (chainId: Chain) => {
+  const chain = getChainByNetworkId(chainId);
+  return new Connection(chain.nodeUri[0]);
```
Comment (maximgeerinck) [inline src/utils/connection.ts:7]: chain can be not found

**Review by maximgeerinck — APPROVED** [review]
Looking rather good for such a PR, small things but nothing too serious i think.
Just find the 2 refinable clients confusing and not that user friendly.

File: src/solana/actions/addTokensToVault.ts (hunk 7 above)
Comment (Superjo149) [reply src/solana/actions/addTokensToVault.ts:61]: This is how its done, if its a potential error, its an errr in 500 places :D

File: src/config/sdk.ts (hunk 6 above)
Comment (Superjo149) [reply src/config/sdk.ts:47]: Woops, testing purposes on local

File: examples/item/itemsOnAuction.ts (hunk 5 above)
Comment (Superjo149) [reply examples/item/itemsOnAuction.ts:12]: This is the case, just not using it in this example. Will adjust

File: src/solana/oyster/models/metaplex/startAuction.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  const value = new StartAuctionArgs();
+  const data = Buffer.from(serialize(SCHEMA, value));
+
+  console.log({
+    auctionManagerKey: auctionManagerKey,
+    auctionKey,
+    auctionManagerAuthority,
+    storePubKey,
+    programId: PROGRAM_IDS.auction,
+  });
```
Comment (deveshrawat) [inline src/solana/oyster/models/metaplex/startAuction.ts:28]: leftover ??

**Review by deveshrawat — APPROVED** [review]
Seems ok

File: src/nft/solana/connection.ts [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+  let breakEarlyObject = { breakEarly: false, i: 0 };
+  console.log(
+    "Signed txns length",
+    signedTxns.length,
+    "vs handed in length",
+    instructionSet.length
+  );
+  const txIds = [];
+  for (let i = 0; i < signedTxns.length; i++) {
```
[the remark this answers is not in the document]
Comment (Superjo149) [reply src/nft/solana/connection.ts:75]: Because they are not supposed to be running parallel

File: src/offer/SaleOffer.ts (hunk 1 above)
Comment (Superjo149) [reply src/offer/SaleOffer.ts:41]: If it fails, it will throw an error

File: src/refinable/account/SolanaAccount.ts (hunk 2 above)
Comment (Superjo149) [reply src/refinable/account/SolanaAccount.ts:21]: Not sure what you mean

File: src/transaction/EvmTransaction.ts (hunk 3 above)
Comment (Superjo149) [reply src/transaction/EvmTransaction.ts:20]: Yes, that the whole idea behind the wait

File: src/utils/connection.ts (hunk 9 above)
Comment (Superjo149) [reply src/utils/connection.ts:7]: The utility method already handles this

File: src/solana/oyster/models/metaplex/startAuction.ts (hunk 10 above)
Comment (Superjo149) [reply src/solana/oyster/models/metaplex/startAuction.ts:28]: Yeah, still quite some logs everywhere

**Pull request merged**